### PR TITLE
Allow applying a new profile over an existing one in ***pan-os-apply-security-profile*** command

### PR DESCRIPTION
**Pull request by rshunim**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-6832

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by rshunim** [comment]
@GuyAfik 
I checked it in the UI for Panorama and PAN-OS.
If you want, we can do a short demo together.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 1]
```diff
@@ -6916,14 +6916,40 @@ def get_security_profiles_command(security_profile: str = None):
...
+    profile_types = {'data-filtering', 'file-blocking', 'spyware', 'url-filtering',
+                     'virus', 'vulnerability', 'wildfire-analysis'} - {profile_type}
+
+    # first we update the given profile type with the given profile name
+    element = f"<{profile_type}><member>{profile_name}</member></{profile_type}>"
+
+    # Keeping the existing profile types
+    for p_type in profile_types:
+        if p_type in profile_types_result:
+            element += f"<{p_type}><member>{profile_types_result[p_type]['member']['#text']}</member></{p_type}>"
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6943]: `['#text']` is only true when the profile is un-committed, it fails if the rule was already committed.

use the `parse_pan_os_response` function to parse un-committed pan-os response to be like committed response

make sure to use `get` of `dict` instead also.

File: Packs/PAN-OS/ReleaseNotes/1_17_5.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### Palo Alto Networks PAN-OS
+
+- Fixed an issue where ***pan-os-apply-security-profile*** command was failing when applying a new profile over an existing one.
```
Comment (GuyAfik) [inline Packs/PAN-OS/ReleaseNotes/1_17_5.md:6]:
```suggestion
- Fixed an issue where the ***pan-os-apply-security-profile*** command was failing when applying a new profile over an existing one.
- Fixed an issue where the ***pan-os-apply-security-profile*** command didn't work for panorama instances.
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 3]
```diff
@@ -6933,20 +6959,18 @@ def apply_security_profile_command(args):
     profile_name = args.get('profile_name')
     profile_type = args.get('profile_type')
     rule_name = args.get('rule_name')
-    pre_post = args.get('rule_name')
+    pre_post = args.get('pre_post')
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6962]: nice catch

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 4]
```diff
@@ -6916,14 +6916,40 @@ def get_security_profiles_command(security_profile: str = None):
...
+        'action': 'get',
+        'type': 'config',
+        'xpath': xpath,
+        'key': API_KEY,
+    }
+    result = http_request(URL, 'GET', params=params)
+
+    # Get all profile types already existing, so we don't override them when updating
+    profile_types_result =\
+        result.get('response', {}).get('result', {}).get('entry', {}).get('profile-setting', {}).get('profiles', {})
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6931]: use maybe `dict_safe_get` from common server python instead

**Review by GuyAfik — CHANGES_REQUESTED** [review]
@rshunim nice work! please write a unit-test to cover this if possible.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 4 above)
Comment (rshunim) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:6931]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 5]
```diff
@@ -6916,14 +6916,44 @@ def get_security_profiles_command(security_profile: str = None):
...
+                     'virus', 'vulnerability', 'wildfire-analysis'} - {profile_type}
+
+    # first we update the given profile type with the given profile name
+    element = f"<{profile_type}><member>{profile_name}</member></{profile_type}>"
+
+    # Keeping the existing profile types
+    for p_type in profile_types:
+        if p_type in profile_types_result:
+            member = profile_types_result.get(p_type, {}).get('member')
+            element += f"<{p_type}><member>{member}</member></{p_type}>"
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6947]:
```suggestion
            p_name = profile_types_result.get(p_type, {}).get('member')
            element += f"<{p_type}><member>{p_name}</member></{p_type}>"
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 6]
```diff
@@ -6916,14 +6916,44 @@ def get_security_profiles_command(security_profile: str = None):
...
+
+    # keep the context aligned with both committed and un-committed objects
+    parse_pan_os_un_committed_data(profile_types_result, ['@admin', '@dirtyId', '@time'])
+
+    # remove from the types the given profile type, since we update it anyway
+    profile_types = {'data-filtering', 'file-blocking', 'spyware', 'url-filtering',
+                     'virus', 'vulnerability', 'wildfire-analysis'} - {profile_type}
+
+    # first we update the given profile type with the given profile name
+    element = f"<{profile_type}><member>{profile_name}</member></{profile_type}>"
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6941]:
```suggestion
    rule_profiles = f"<{profile_type}><member>{profile_name}</member></{profile_type}>"
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 7]
```diff
@@ -6916,14 +6916,44 @@ def get_security_profiles_command(security_profile: str = None):
...
+            member = profile_types_result.get(p_type, {}).get('member')
+            element += f"<{p_type}><member>{member}</member></{p_type}>"
+
     params = {
         'action': 'set',
         'type': 'config',
         'xpath': xpath,
         'key': API_KEY,
-        'element': f'<member>{profile_name}</member>'
+        'element': f'<profile-setting><profiles>{element}</profiles></profile-setting>'
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6954]:
```suggestion
        'element': f'<profile-setting><profiles>{rule_profiles}</profiles></profile-setting>'
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 8]
```diff
@@ -6916,14 +6916,44 @@ def get_security_profiles_command(security_profile: str = None):
...
-def apply_security_profile(xpath: str, profile_name: str) -> Dict:
+def apply_security_profile(xpath: str, profile_name: str, profile_type: str) -> Dict:
+    # first api call just for getting the current status
+    params = {
+        'action': 'get',
+        'type': 'config',
+        'xpath': xpath,
+        'key': API_KEY,
+    }
+    result = http_request(URL, 'GET', params=params)
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6927]:
```suggestion
    # get the rule state
    params = {
        'action': 'get',
        'type': 'config',
        'xpath': xpath,
        'key': API_KEY,
    }
    result = http_request(URL, 'GET', params=params)
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 9]
```diff
@@ -6916,14 +6916,44 @@ def get_security_profiles_command(security_profile: str = None):
...
+
+    # Get all profile types already existing, so we don't override them when updating
+    profile_types_result = dict_safe_get(result, ['response', 'result', 'entry', 'profile-setting', 'profiles'],
+                                         default_return_value={})
+
+    # keep the context aligned with both committed and un-committed objects
```
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6933]:
```suggestion
    # align the response for both committed and un-committed profiles 
```

this isn't needed for the context as we dont return context in this command.

**Review by GuyAfik — APPROVED** [review]
looks great, few minor comments, fix them and merge

**Pull request merged**